%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to 
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%         <14 August 2007>            %%
%%                                     %%
%%                                     %%
%% Uses:                               %%
%% cite.sty, url.sty, bmc_article.cls  %%
%% ifthen.sty. multicol.sty		   %%
%%				      	   %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%	
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.pdf and the instructions for    %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}    



% Load packages
\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
\usepackage{graphicx}
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
 
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%   
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %% 
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %% 
%%  submitted article.                         %% 
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                     


%\def\includegraphic{}
%\def\includegraphics{}



\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                              %%
%% You may change the following style settings  %%
%% Should you wish to format your article       %%
%% in a publication style for printing out and  %%
%% sharing with colleagues, but ensure that     %%
%% before submitting to BMC that the style is   %%
%% returned to the Review style setting.        %%
%%                                              %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 

%Review style settings
%\newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
%\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

%New style setting
\newenvironment{bmcformat}{\baselineskip20pt\sloppy\setboolean{publ}{false}}{\baselineskip20pt\sloppy}



% Begin ...
\begin{document}
\begin{bmcformat}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{BiNoM, a Cytoscape plugin for accessing and analyzing pathways using
standard systems biology formats}
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Ensure \and is entered between all but   %%
%% the last two authors. This will be       %%
%% replaced by a comma in the final article %%
%%                                          %%
%% Ensure there are no trailing spaces at   %% 
%% the ends of the lines                    %%     	
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% \author{Jane E Doe\correspondingauthor$^1$%
%          \email{Jane E Doe\correspondingauthor - jane.e.doe@cambridge.co.uk}
%        and 
%          John RS Smith$^2$%
%          \email{John RS Smith - john.RS.Smith@cambridge.co.uk}%
%       }
%       

\author{Eric Bonnet$^{1,2,3}$ \and Laurence Calzone$^{1,2,3}$ \and Daniel
Rovera$^{1,2,3}$ \and Gautier Stoll$^{1,2,3}$ Emmanuel Barillot$^{1,2,3}$ and Andrei Zinovyev$^{1,2,3}$\correspondingauthor
\email{andrei.zinovyev@curie.fr} }
       





%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address{%
    \iid(1)Institut Curie, 26 rue d'Ulm, Paris, F-75248 France\\
    \iid(2)INSERM, U900, Paris, F-75248 France\\
    \iid(3)Mines ParisTech, Fontainebleau, F-77300 France
}
\maketitle

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%  
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%   
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\begin{abstract}
        
Public repositories of biological pathways and networks have greatly expanded in
recent years. Such databases contain many pathways that facilitate the analysis of high-throughput
experimental work and the formulation of new biological hypotheses to be
 tested, a fundamental principle of the Systems Biology approach.
However, large scale molecular maps are not always easy to mine and interpret. We
have developped BiNoM (Biological Network Manager), a Cytoscape plugin which provides
 several functions for the import-export of standard Systems Biology
file formats, and also a set of functions to analyze and reduce the complexity
of large biological networks. Here, we provide an in-depth overview of the BiNoM
functions, we detail novel aspects such as the support of the BioPAX level 3
format and an algorithm for the quantification of pathways for influence
networks. We also illustrate some of the BiNoM functions on a detailed biological case
study of a network representing the G1/S transition of the cell cycle.
        
\end{abstract}



\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}




%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %% 
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%




%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Background}

Biological pathways and networks represent sets of interactions, or functional
relationships, occurring at the molecular level in living cells
\cite{adriaens2008public, cary2005pathway}. Such representations represent now
a large body of knowledge that are often organized in publicly available
repositories such as the KEGG database \cite{ogata1999kegg}, Reactome
\cite{joshi2005reactome}, MINT \cite{zanzoni2002mint}, or the Cancer Cell Map
(\url{http://cancer.cellmap.org/cellmap/}). All these pathway and biological
network databases facilitate a large spectrum of analyses, improving our
understanding of cellular systems. For example, it is now a very common
practise to cross the output of high-throughput experiments, such as mRNA or
protein expression levels, with with curated biological pathways in
order to efficiently visualize changes, analyze their impect on a network and
formulate new hypotheses about
biological processes \cite{saraiya2005visualizing,
gehlenborg2010visualization}. The development of those pathway repositories has
also fueled the creation of standard representations and formats, to facilitate
the exchange and representation of data, such as the Biological Pathway
Exchange standard (BioPAX) \cite{demir2010biopax}, the Systems Biology Markup
Language (SBML) \cite{hucka2003systems} or the Systems Biology Graphical
Notation (SBGN) \cite{le2009systems}. The Pathway resource list website counts
more than 300 web-accessible biological pathway and network databases
\cite{bader2006pathguide}, many of which are using the SBML and BioPAX standard
formats. Ultimately, those integrated resources will facilitate computational
models building and their experimental validation, a cycle that is the
cornerstone of the Systems Biology initiative \cite{karlebach2008modelling,
kitano2002systems, ideker2001new}.

As a consequence, there is a need for the precise and accurate reconstruction of
pathways and large scale molecular maps covering fundamental biological
processes. Such maps are often constructed by manual curation of literature or
automated curation from pathway databases \cite{bauer2009pathway}. They also
often focus on the regulation of biological processes involved in diseases such
as cancer \cite{oda2005comprehensive, oda2006comprehensive,
calzone2008comprehensive, caron2010comprehensive}. However, the scale of such
maps, even when they are focusing on a particular process, is quite large, with
hundreds of species and interactions. The analysis and interpretation of such
large scale maps is therefore not a straightforward task. Several computational
tools have been developed to facilitate the visualization, curation and analysis
of pathways \cite{adriaens2008public}. For example, CellDesigner is a software package
for the graphical edition of biological pathways diagrams
\cite{funahashi2003celldesigner}, and it is using a dialect
of SBML to store all the information contained in the graphs. There is obviously
a need for user-friendly software tools that would allow the user to easily
import data from various standard format sources, to perform structural analyses
and to manipulate networks, and also be able to export a network to a suitable
format for further analysis (e.g. mathematical modeling). We have created
BiNoM \cite{zinovyev2008binom}, a software plugin for the popular Cytoscape
network vizualization and analysis tool \cite{cline2007integration}. BiNoM
facilitates the usage and analysis of biological networks in standard Systems
Biology formats (SBML, SBGN, BioPAX). More specifically, BiNoM implements
several functions based on graph operations for the structural analysis of
biological networks. Those functions can be used to reduce the complexity and
extract meaningful subnetworks from large scale molecular maps. Furthermore,
BiNoM has several built-in functions for accessing standard Systems Biology file
formats. 
Here, we provide a detailed view on the functions implemented in BiNoM for the
analysis of large scale molecular maps. We also highlight novel functions that
were implemented recently, such as the support of the latest BioPAX
specification (BioPAX level 3) and an algorithmic approach for the
quantification of pathways on influence networks (PIQuant). We illustrate the
usage of the principal BiNoM functions with a detailed analysis of a molecular network
representing the G1/S transition of the cell cycle, a central mechanism for
tumor development and progression.

\section*{Implementation}
The BiNoM software is implemented in the Java programming language, as a plugin
for the network visualization and analysis software package Cytoscape
\cite{cline2007integration}. Although the primary use of BiNoM is through the
Cytoscape software, the underlying logic of all BiNoM functions is completely
decoupled from the Cytoscape objects, allowing developers to also use BiNoM as
an independent java library \cite{zinovyev2008binom}. The installation of BiNoM
can be done through the Cytoscape plugin manager, or alternatively the user can
also download the plugin together with a manual and the source code from the
BiNoM website
\url{http://binom.curie.fr/}. A detailed tutorial on how to use the main BiNoM
functions is in press
\cite{binomBookChapter}.


BiNoM is designed to handle different Systems Biology file formats and to
provide useful functions for the analysis of biological networks. The core
functions of BiNoM can be grouped in five different topics: Input/Output,
Analysis, BioPAX utils \& query, Module manager and Utilities.


\subsection*{BiNoM Input/Output}

BiNoM functions facilitate the import and export of standard Systems Biology
file formats, but is not designed to be a universal converter. The goal is more
to provide robust functions for a limited number of useful conversion
scenarios: 

\begin{itemize}

\item Interconversion of CellDesigner files to BioPAX, and from a BioPAX
reaction network to SBML level 2.

\item Import of a BioPAX file as a reaction network and/or a pathway structure
and/or a protein interactions network, followed by the creation of a subnetwork
saved as a new BioPAX file. 

\item Import of a BioPAX file, selection of a subnetwork of interest saved as a SBML file for
the creation of a computational model using an appropriate software package.

\item Import a large CellDesigner map and export only a subnetwork as a new CellDesigner file.  

\end{itemize}

BiNoM allows the user to import and export SBML level 2 files, as well as
CellDesigner 3.x and 4.x file formats \cite{zinovyev2008binom}. The BioPAX
community has recently made a major update of the BioPAX standard, producing a
new specification known as BioPAX Level 3 (\url{http://www.biopax.org/}). This
format supports metabolic pathways, signaling pathways (including states of molecules
and generic molecules), gene regulatory networks, molecular interactions and
genetic interactions. Due to major changes in the
specification, the BioPAX level 3 is incompatible with the level 2 file format.

BioPAX is using the web ontology language specification (OWL,
\url{http://www.w3.org/2004/OWL/}) to store data in XML-formatted files. In
BiNoM, we use the Jastor and the Jena java libraries
(\url{http://jastor.sourceforge.net/}, \url{http://jena.sourceforge.net/}) to
automatically create java classes from the BioPAX specifications, allowing a
convenient access to the different data types encoded in the BioPAX files. The
guiding principle in BiNoM for the
access to Systems Biology files is to provide control over the content without
completely converting the file to the Cytoscape format. The content is therefore
mapped to a labeled directed graph, representing the complete set of objects and
their relationships. This graph, called the index, is highly connected and is
not visualized explicitely. Instead, the user is interacting with selected subgraphs
extracted from the index. For example, BioPAX files are imported as three
separate entities, respectively the Reaction Network (RN), representing the
biochemical reaction network, the Pathway Structure (PS), showing the
hierarchical arborescence of pathways, and the Protein-Protein interaction graph
(PP). An example of a reaction network imported from a BioPAX level 3 file is
shown on figure~\ref{mphasebiopax}, while the figure~\ref{apoptosishierarchical}
is showing the pathway hierarchical
structure extracted from the human Apoptosis pathway.  


\begin{figure}[h]
 \caption{\label{mphasebiopax}  \textbf{Cell cycle network.} Zoom on the cell
cycle network model of Novak et al. \cite{novak1998model} after the import in 
Cytoscape of the corresponding BioPAX file with BiNoM functions. The grey
squares represent complexes while white squares are proteins. Small grey diamonds
indicate reactions. The names of each node follow BiNoM naming convention (see the main text for details)}
 \center{\includegraphics[width=0.7\textwidth]{figures/mphase_biopax.png}}
\end{figure}

\begin{figure}[h]
 \caption{\label{apoptosishierarchical}  \textbf{Apoptosis pathway hierarchical
structure.} Zoom on the BioPAX data extracted from the Reactome database
\cite{joshi2005reactome}. The green nodes represent pathways, the pink
triangular nodes denote steps, while grey nodes indicate reactions.}
 \center{\includegraphics[width=0.7\textwidth]{figures/apoptosis_hierarchical.png}}
\end{figure}


When importing a file, BiNoM is calling a \emph{naming service} function, in
order to create meaningful names for the various entities. More precisely,
entity names are combined with other features such as modifications, compartment
and complexe components. The different features are indicated by special
characters, such as ``@'' for the compartments, ``$|$'' for modifications and
``:'' to delimitate the different members of a complex. For example, the
name Cdc25$|$Pho@cytoplasm represents the protein Cdc25 in a phosphorylated
state, located in the cytoplasm, while the name
Cdc13:Cdc2$|$Thr167\_pho@cytoplasm indicates a protein complex located in the
cytosplasm, composed of the protein Cdc13 and the protein Cdc2 phosphorylated at
position 167 on a threonin residue.


\subsection*{BiNoM Analysis}
The central goal of the BiNoM plugin is to provide efficient methods
and algorithms to reduce the inherent complexity of biological networks into
manageable and meaningful subnetworks. This goal is achieved by a set of
functions included as a built-in structural graph analysis library. Some of the functions are taking into
account the semantics contained in the graph elements names. 
The structural analysis functions implemented in BiNoM include the analysis of
connected and strongly connected components, pruning the network, decomposition
by name (material components) or by cycle, path analysis and network clustering.


\subsubsection*{Decomposition by name or by cycle}
A trivial approach to separate a network into subparts is to dissociate the unconnected subparts
of the network. A more sophisticated one consists in decomposing the network into
strongly connected components, using the algorithm of Tarjan
\cite{tarjan1972depth}. It is also possible to prune the network into three different parts:
the one with all the elements associated with the \emph{input} parts of the network, the
second with all the elements associated with the \emph{output} part and the last
part with all the elements linked to the central, cyclic part. This type of approaches
corresponds to the bow-tie model of Broder and colleagues \cite{broder2000graph}.

The decomposition in material components is using the node name semantics to
isolate subnetworks in which each protein is involved. Major overlaps between
the different subnetworks are to be expected, as some proteins are involved in
different complexes. Figure~\ref{matcdc2cdc13} shows two examples of subnetworks
obtained by material component decomposition applied to the cell cycle network
model.

\begin{figure}[h]
 \caption{\label{matcdc2cdc13}  \textbf{Subnetworks obtained after a decomposition in
material components.} The two overlapping subnetworks correspond to the components Cdc13 and
Cdc2.}
 \center{\includegraphics[width=0.7\textwidth]{figures/mat_cdc2_cdc13.png}}
\end{figure}


The cycle decomposition is splitting the network into relevant directed cycles
\cite{gleiss2001relevant}, using a modifed version of the algorithm of Vismara
and colleagues \cite{vismara1997union}. Very often, this approach will represent
the different parts of the life cycle of a given protein or complex. Care must
be taken when applying this approach, as the number of cycles can be huge for
large network structures. For example, it might be preferable to eliminate first
the network hubs, which are by definition highly connected, and also group short
cycles in larger subnetworks before applying the procedure. Figure~\ref{mphasecdc25cycles} shows two
cycles involving CDC25 after a cycle decomposition.

\begin{figure}[h]
 \caption{\label{mphasecdc25cycles}  \textbf{CDC25 subnetworks.}
      Two cycles for the CDC25 protein found after the decomposition of the cell cycle network model of Novak et al. \cite{novak1998model}.}
 \center{\includegraphics[width=0.7\textwidth]{figures/mphase_cdc25_cycles.png}}
\end{figure}




\subsubsection*{Path analysis algorithms}

[[ AZ: check description of shortest path and sub-optimal path search and add a description for all non intersecting paths ]]

BiNoM analysis functions also include several classical path analysis
algorithms, such as finding the shortest paths, the suboptimal shortest paths or
all non-intersecting paths. The shortest path is calculated as the path having
the minimal sum of weights of the edges composing the path (Dijkstra's
algorithm) while the sub-optimal path is constructed by removing all edges in
all shortest paths one by one, and finding the shortest path. The user should
be careful when using this procedure, as the number of paths between nodes can be
very large for big networks.

Obviously, the result of some decomposition functions will result in subnetworks
that share a large number of components, as it is for example often the case
with the decomposition in material components. Therefore, BiNoM also includes a
function to \emph{cluster} networks, based on common components such as protein
or protein complexes. To determine the size of the clusters, the user can
specify a percentage of intersection (ranging from 0 to 100\%) that will be used
as a threshold to create the clusters.

\subsubsection*{Pathway influence quantification algorithm}
In this version of BiNoM, we have introduced a novel approach to quantify the
influence of one or more source nodes on one or more target nodes for a given network, 
named the PIQuant score (Pathway Influence Quantification). For example,
Figure~\ref{piquantnetworks}a shows a simple influence network composed of seven
nodes, where the edges represent either a repression or an activation between
nodes. The node Ac represents an input node of interest, while the node Ph is
defined as an output node. The latter can be a gene or a phenotype of interest,
representing a more complex biological function, such as cell proliferation or
apopotosis. Figures~\ref{piquantnetworks}b and c show two examples of paths defined between
the nodes Ac and Ph. In this case, the PIQuant score value will quantify the influence of the 
source node Ac on the target node Ph, taking into account available experimental
data values for input nodes, paths length and the sign of the paths. More generally, the goal
of the PIQuant score is to help the user to understand the role and function of
the input nodes on the output nodes, for a given network architecture in some defined experimental conditions.

More formally, we define a source node as \textit{annotated} when a signed real number is
assigned to the node, representing experimental data. For example, it can be the
expression ratio of a gene between a disease and a normal state, obtained from
transcriptomics profiling.
We can also extract a set of paths from source nodes to target nodes (indexed
from $1$ to $q$). 
 A path $k \in \{1,\ldots , q\}$ is defined
as a the sequence of consecutive connected nodes
between a source node and a target node (without repetition of any node
or edge). In BiNoM, we propose three algorithms to search for paths between the
source and target nodes (shortest paths, sub-optimal shortest paths and all
non-intersecting paths, see previous paragraph). The activity
$\alpha_k$ of the path $k$ is defined as the annotation
of its source node. We define the sign $\sigma_k$
of the path $k$ as the product of the signs of every edge of the path and finally
the length $\lambda_k$ of the path $k$ as the number of edges in the path. We also
hypothesize that the longer the path is, the lesser the global influence will be
on the target node. Then the PIQuant score for a set of $q$ paths (from a set of 
source nodes, to a set of target nodes) is defined as:

$$
 PIQuant_{Score} = \sum_{k=1}^{q} \alpha_{k} \sigma_{k} \frac{1}{\lambda_{k}}
$$

For example, in the case of the network figure~\ref{piquantnetworks}a, let us
consider Ac as the source node and Ph as target node and consider only the two
paths defined on the figures~\ref{piquantnetworks}b and
~\ref{piquantnetworks}c. Given that the
node Ac is annotated by the value $2.0$, and that the first path has a length
equal to 3, the second a length equal to 5, we can calculate the PIQuant score 
of the node Ac to the node Ph as:

$$
 PIQuant_{Score} = 2 \cdot 1 \cdot \frac{1}{3} + 2 \cdot (-1) \cdot \frac{1}{5}
= 0.27
$$


Sometimes a path has one or more intermediate nodes that are annotated nodes, 
in addition to the source node.
In that case, the intermediate node is \textit{consistent} if the sign of its 
annotation is the same as the sign of the source node annotation multiplied by
the sign of the 
path from the source node to the intermediate node. If signs are opposite, the
node is \textit{inconsistent}. 
A path is \textit{consistent} if each annotated intermediate nodes is
consistent. A path is \textit{inconsistent} if at least one intermediate node is
\textit{inconsistent}. 

The rationale behind these definitions is that the
network structure for intermediate nodes can be consistent or inconsistent with
experimental data values.

For example, the path shown on figure~\ref{piquantnetworks}b is consistent while
the path on the figure~\ref{piquantnetworks}c is inconsistent. In the example
above, if only consistent paths are kept, then the PIQuant score of Ac to Ph becomes:

$$
 PIQuant_{Score} = 2 \cdot 1 \cdot \frac{1}{3} = 0.67
$$

According to the network, there is an influence of Ac to Ph
that goes through D (Figuree~\ref{piquantnetworks}c). But if this path
is functional, D annotation would have the same sign than Ac annotation, because
Ac activates D indirectly. In our case, the sign of D annotation is opposite to
the sign of Ac annotation, meaning that there is another (unknown) effect on D.
Therefore, the path from Ac to Ph that goes through D is \textit{inconsistent} and
might not be used to compute PIQuant score. Practically, we offer the option to
keep or not the inconsistent paths for the calculation of the PIQuant score,
depending on how the user wants to analyze the results.

In more realistic situations, we will have multiple source nodes with different
 annotations, and also multiple target nodes representing phenotypes of
particular nodes of interest. We have implemented in BiNoM a set of functions
that allow users to select source nodes, select target nodes, and choose among
three different options for searching paths (shortest paths, optimal and
suboptimal shortest paths, all non-intersecting paths). The software is then
calculating PIQuant scores for every target node specified, taking into account
every possible path found by the algorithm. We have also implemented a
randomization test to estimate p-values for each path. We describe a detailed
and concrete application of the PIQuant algorithm to a real biological network
in the results section.

\begin{figure}[h]
 \caption{\label{piquantnetworks}  \textbf{A simple influence network.} The 
network is composed of seven nodes and nine edges (a). The two paths (b,c)
extracted from this network start from the source node \textbf{Ac} and end at
the target node \textbf{Ph} (which denotes a phenotype of interest). The nodes
\textbf{Ac} and \textbf{D} are annotated using experimental data, having the
values $2.0$ and $-1.0$ respectively.}
 \center{\includegraphics[width=0.7\textwidth]{figures/piquant_networks.png}}
\end{figure}

\subsection*{BiNoM BioPAX utils \& query}
The BioPAX format was primarily conceived as a standard facilitating the
exchange of data between various database systems \cite{demir2010biopax}. As a
consequence, this format was designed first to be machine-readable, but was not
intended
to be edited and modified by biologists. Furthermore, due to its adoption by
large biological knowledge repositories, some BioPAX files can be really big,
such as the \textit{Homo sapiens} network from the reactome database
\cite{joshi2005reactome} that has more than 6,000 reactions involving more than
8,000 different species (proteins, RNA molecules, metabolites, etc.).

BiNoM implements a set of functions precisely aiming at allowing end users to
easily visualize and modify BioPAX files. To the best of our knowledge, there is
no other software package allowing to view, query and edit BioPAX files than
BiNoM at the moment. The functions are using
Java class introspection techniques to build a BioPAX class tree. Then, the
content of the file can easily be accessed. For example, figure~\ref{biopaxtrailprop} shows
all the information linked to the TRAIL protein, after a call to the
BioPAX property
editor function of the BiNoM menu. Details are given in the BiNoM manual on how
to display valid attributes and edit them, and also how to visualize the
complete BioPAX tree.


\begin{figure}[h]
 \caption{\label{biopaxtrailprop}  \textbf{Extra Information linked to the TRAIL protein.}
      The informations are automatically extracted from the BioPAX file upon
import with the BiNoM I/O functions.}
 \center{\includegraphics[width=0.7\textwidth]{figures/biopax_trail_prop.png}}
\end{figure}


The BioPAX Query functions in BiNoM allow the user to work with potentially huge
BioPAX data files and extract the relevant information, by querying an index and
retrieving data from it. The index corresponds to a mapping of the content of
the BioPAX file on a labeled graph (an index file is created and saved, using
the XGMML format). Various statistics can be displayed on the content of the
index, such as the number of proteins, complexes, reactions, publications, etc.
To start extracting relevant information, the user can query the index by gene
name (and/or by any synonym of the gene) and start building a network centered
around this molecule of interest. The extension of the network is done by adding
different types of entities: complexes where the molecule of interest is
involved, chemical species, reactions (with the possibility of including all the
sources and targets of the reactions) and related publications. Figure~\ref{smaccomplexes}
illustrates an example of a small network extracted from the human apoptosis
pathway downloaded from the Reactome database \cite{joshi2005reactome}, and
centered on the SMAC protein, with all the protein complexes in which
this protein is involved and that were added using the BiNoM BioPAX Query functions.


\begin{figure}[h]
 \caption{\label{smaccomplexes}  \textbf{SMAC pathway subnetwork.}
      Subnetwork extracted from the human Apoptosis pathway, starting with the
SMAC protein and expanding to all protein complexes where this
molecule is involved using the BiNoM Query functions.}
 \center{\includegraphics[width=0.7\textwidth]{figures/smac_complexes.png}}
\end{figure}


\subsection*{BiNoM Module manager}
The application of selected BiNoM analysis algorithms to a large network map,
usually followed by manual curation, results in the creation of a certain number
of connected network \emph{modules}. Most of the modules represent a detailed
sequence of events that occur with a particular protein or protein complex,
whose name can then be used to represent the whole module. This way, a
simplified representation of a complex map can be produced, using the modules
and their relationships as an abstracted version of the comprehensive network
\cite{calzone2008comprehensive}.

To facilitate the creation and management of modules, we have used in this
version of BiNoM a new feature introduced in recent versions of Cytoscape (as of
version 2.7) \cite{cline2007integration}, known as \emph{nested} networks. This
feature allows to embed any cytoscape network in a (meta)node. It was
introduced to allow the creation of network hierarchies and circular
relationships. In BiNoM, we use this feature to facilitate the process of
modularization of a large network. The BiNoM module manager integrates a set of
funtions that allow to easily create a network from a list of subnetworks,
packing individual nodes, merging different subnetworks, displaying informations
about metanodes and calculating intersection between subnetworks.

\subsection*{BiNoM Utilities}
This set of functions correspond to various small utilities that are not
implemented in Cytoscape yet, but might be very useful for the analysis and
manipulation of networks. For example, it is possible to automatically select
the edges between two nodes, generate the network corresponding to the double
network differences between two networks A and B, or update all the subnetworks of
a session after some changes have been made to the initial one. The BiNoM utilities also implements
clipboard related functions, giving the possibility to copy, add and paste
selected nodes and edges and also to show the clipboard content. 


\section*{Results and Discussion}
We propose to study a reaction network focusing on the transition from G1 phase
(growth phase) to S phase (DNA replication phase) of the cell cycle
\cite{calzone2008comprehensive} as an example of the use of BiNoM functions. 

The network published in \cite{calzone2008comprehensive} describes in
biochemical details the regulation of the well-known and charaterized tumor
suppressor gene retinoblastoma (RB or RB1). The product of this gene operates at
the heart of the cell cycle clock, acting as a signal transducer connecting the
cell cycle clock with the transcriptional machinery. The pathway in which RB is
acting is disrupted in many human tumor types \cite{weinberg1995retinoblastoma}.
The comprehensive map of the RB/E2F network was built using CellDesigner
\cite{funahashi2003celldesigner} and listed 80 proteins, 208 species, 165
interactions, 176 genes, and recapitulated more than 350 publications, including
information from different cellular types, thus making the map a generic map of
the cell cycle regulation. It is composed of two main compartments: the cell,
containing the cytoplasm, the nucleus and the nucleolus, in which the
biochemical interactions such as association, dissociation, (de)phosphorylation,
(de)acetylation, degradation, etc. take place; and the genes, which lists the
target genes of the main transcription factors of the map, the E2F family
members.
A thorough description of the model and the methods used to build it and create
simplified versions of it along with the first interactive (clickable) map is
available on our website
(\url{http://bioinfo-out.curie.fr/projects/rbpathway/interactive/rb_network.html/}).

For this study, we choose to concentrate on the G1 to S transition. We
used the intersection of the 208 species
of RB/E2F network and the 280 species listed in Reactome \cite{joshi2005reactome} for the G1-S transition
(referred to as \emph{Mitotic G1 G1/S phases} in Reactome). The resulting
subnetwork contains 38 proteins, 98 chemical species, and 100 biochemical
reactions (Figure~\ref{g1s}).

\begin{figure}[h]
 \caption{\label{g1s}  \textbf{G1/S network.}
	Overview of the G1 to S transition network, corresponding to the intersection of
RB/E2F network and G1/S network extracted from Reactome.}
 \center{\includegraphics[width=0.85\textwidth]{figures/G1S.png}}
\end{figure}

This map contains a lot of valuable informations but they are rather difficult to extract. We present two
ways to get some biological insight from this map. The first one consists of
transforming the reaction network into an influence network in order to analyze
experimental data on it. The second one implies a simplification of the comprehensive map by
applying a method of reduction of the numerous interactions into modules without
losing any content from the original map.

\subsection*{Application of PIQuant on an influence network.}
PIQuant algorithm can be used to perform a quantitative analysis on an influence
network. For translating the reaction network into an influence network, we used
a tool developed by the team of BIOCHAM \cite{calzone2006biocham} that is
available online (\url{http://contraintes.inria.fr/~soliman/cd2dot.html}). The
translation is based on simple rules: (1) BIOCHAM deletes all linear
degradations and syntheses, (2) all intermediary species with only one input and
one output, are suppressed, (3) if the reactions of synthesis and degradation of
a species have distinct inputs and outputs, then this reactions can be merged,
and (4) if the inputs of one reaction is the same as the inputs of another
reaction, then it is the case of a reversible reaction. Although this work is
still in progress, a short description and an example of this conversion is available \cite{calzone2011calamar}. 

We apply PIQuant to the resulting influence network of the G1/S transition of
the cell cycle. 

In order to construct the annotation of this network, we use expression  data
from a study of 57 bladder cancer
tissue samples compared to normal tissue \cite{stransky2006regional}. For each 
gene, the differential expression between tumor and normal tissue
is assessed by a t-test. The t-test statistic value is used as the annotation
for each gene. We construct a text file listing nodes of the influence network 
and their annotation. We assign annotations only to nodes that can be directly
influenced by a 
mRNA level. When a complex is annotated, it means that one of the partners has
no unbound state in the network. 
We import this file using the Cytoscape function 
``Import $>$ Nodes attributes'' 
in the cytoscape session of the influence network.
Figure~\ref{InflAnnotNet} represents this influence network with the annotation.
 

\begin{figure}[h]
  \caption{\label{InflAnnotNet} \textbf{Annotated influence network.}
    Influence network of the cell cycle G1/S transition. Colors represent differential
expression obtained from transcriptome data compared to normal tissue. Color 
intensities are proportional to the t-statistics (red for positive, green for
negative). The three grey nodes are markers for the different cell-cycle phases: G1, S, M.}
  \center{\includegraphics[width=0.85\textwidth]{figures/Cell_cycle_Expr.pdf}}
\end{figure}

PIQuant is applied to this network and its annotation, by using the function
``Plugins $>$ BiNoM 2.0 $>$ BiNoM Analysis $>$ Path Influence Quantification
analysis''. We take all annotated nodes as active nodes, the grey nodes of
Figure~\ref{InflAnnotNet} as targets, and we select optimal and sub-optimal 
shortest path. We run a test significance and get the report table. In this
report, the ``Node influence table'' represents the global PIQuant score from
each annotated nodes to each cell-cycle phase markers. Those values are
represented as a heat map in Figure~\ref{PIQuantHeatMap}. We can see that most
genes are over-expressed relative to the normal cells in the M and S phases (red
coloring on the heatmap), indicating an enhanced proliferation for tumor cells.

\begin{figure}[h]
  \caption{\label{PIQuantHeatMap} \textbf{Heat map representation of the PIQuant
scores.}
    The map shows the results of the PIQuant algorithm applied to the G1/S
influence network. Color intensities are proportional to PIQuant score (red color indicates a positive
value; i.e an activation, the green color indicates a negative value, i.e. an
inhibition).
Each line represents a source node while each column represents a cell-cycle
phase.}
 
\center{\includegraphics[width=0.8\textwidth]{figures/heatmapPathInfluence.pdf}}
\end{figure}



\subsection*{Modularization} 

The raw G1/S network is very detailed and may be hard to read at a first glance.
To facilitate the analysis of the content, we propose to organise the reaction
network as a modular network. The species are clustered in groups, referred to
as modules, in an semi-automatic manner, using BiNoM functions and biological
knowledge. Each module represent in fact a
sequence of events occuring with a particular protein. The modules are then
linked by activating or inhibiting influences according to the informations
contained in the original diagram or derived from previous biological knowledge.

Briefly, we first decomposed the global network into its different components,
by using name semantics to isolate the sub-networks in which
each protein is involved. The 36 networks that are created this way may share a
lot of common species, so we went further
by clustering the subnetworks having at least 25\% of common species. We renamed
the 7 clusters obtained with a name that illustrate the content and the main
function of the clusters (such as E2F1\_RB, Wee1, etc.). Then, we checked the
content of each module, making modifications if necessary. For example, the
module E2F1\_RB is
further decomposed in three different modules containing the proteins RB, E2F1
and E2F6. Finally, we created a final modular view, where we integrate the
subnetworks we have previously created using a set of BiNoM functions for the
management of modules. Our final modular view is composed of 9 modules, with 22
edges connecting them (Figure~\ref{g1smodular}). A detailed tutorial on the
construction of
this modular network using BiNoM is described in the supplementary methods.


\begin{figure}[h]
 \caption{\label{g1smodular}  \textbf{Modular view of the G1/S network.}
	Modular representation of the G1/S network, created by using a set of
different BiNoM functions. Each node (pictured as a green octagon), represents a
different module, or subnetwork. The edges connecting the modules represent the
 known influences between modules. }
 \center{\includegraphics[width=0.7\textwidth]{figures/G1S_modular.png}}
\end{figure}

The modular view offers a simplified visualization of the complex network. The
obtained model is more abstract but highlights some aspects that may not be
evident from the reaction network. For instance, it brings into relief some
feedbacks, positive, negative or feedforward, involving the major players of the
cell cycle, and prepares the network for mathematical modeling. The translation
of this modular network into a Boolean model is indeed straightforward. 

\section*{Conclusions}

Building a suitable model for systems and mathematical biology is a multi-step
process, beginning with the collection of biological knowledge and progressing
towards the formalization of a network and its translation in mathematical
terms. BiNoM was designed to help during the intermediate steps of this process,
by providing a convenient access to some standard Systems Biology
representations such as BioPAX, also by giving the possibility to manipulate the
network and finally to apply various algorithms (mostly based on graph theory)
and map biological data to it. BiNoM is clearly not a tool for numerical
simulations, but it is providing functions to export networks to the SBML and
GINsim file formats (through the GINsim Cytoscape plugin), facilitating the
import into various numerical simulators.

The future developments of BiNoM will include functions such as merging several
independent networks, finding minimal intervention sets to disrupt or modify the
signaling flow from a set of source nodes to a set of target nodes, ability to
generate a code for web-based representations of biological networks using the
Google Map API and semantic zoom.


\section*{Availability and requirements}

\begin{itemize}
\item Project name: BiNoM
\item Project home page: \url{http://binom.curie.fr/}
\item Operating system(s): Platform independent
\item Programming language: java
\item Other requirements: java 1.5 or higher, Cytoscape 2.7 or higher
\item License: GNU LGPL
\item Any restriction to use by non-academics: none
\end{itemize}




% \section*{Section title}
% \subsection*{Sub-heading for section}
% \subsubsection*{Sub-sub heading for section}
% \subsubsection*{Sub-sub-sub heading for section}

\bigskip

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\section*{Author's contributions}
%    Text for this section \ldots

    

%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Acknowledgements}
  \ifthenelse{\boolean{publ}}{\small}{}
  EB, LC, DR, GS, EmB and AZ are members of the team ‘‘Computational Systems Biology of
Cancer,’’ Equipe labellisée par la Ligue Nationale Contre le Cancer.
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%              
%%  Bmc_article.bst  will be used to                       %%
%%  create a .BBL file for submission, which includes      %%
%%  XML structured for BMC.                                %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %% 
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %% 
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\newpage
{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}
 \bibliographystyle{bmc_article}  % Style BST file
  \bibliography{binom2} }     % Bibliography file (usually '*.bib' ) 

%%%%%%%%%%%

\ifthenelse{\boolean{publ}}{\end{multicols}}{}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

%\section*{Figures}
%  \subsection*{Figure x - title}
%      Description.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
%\section*{Tables}
%  \subsection*{Table 1 - Sample table title}
%    Here is an example of a \emph{small} table in \LaTeX\ using  
%    \verb|\tabular{...}|. This is where the description of the table 
%    should go. \par \mbox{}
%    \par
%    \mbox{
%      \begin{tabular}{|c|c|c|}
%        \hline \multicolumn{3}{|c|}{My Table}\\ \hline
%        A1 & B2  & C3 \\ \hline
%        A2 & ... & .. \\ \hline
%        A3 & ..  & .  \\ \hline
%      \end{tabular}
%      }



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Additional Files}
  \subsection*{Supplementary methods}
	Detailed tutorial for the creation of a modular view of the G1/S network using BiNoM functions.

\end{bmcformat}
\end{document}







