\documentclass[a4,center,fleqn]{NAR}

% Enter dates of publication
\copyrightyear{2008}
\pubdate{31 July 2009}
\pubyear{2009}
\jvolume{37}
\jissue{12}

%\articlesubtype{This is the article type (optional)}


\begin{document}

\title{NaviCell Web Service for Network-based Data Visualization}


\author{%
Eric Bonnet\,$^{1,2,3}$,
Eric Viara\,$^{4}$,
Inna Kuperstein\,$^{1,2,3}$,
Laurence Calzone\,$^{1,2,3}$,
David PA Cohen\,$^{1,2,3}$,
Emmanuel Barillot\,$^{1,2,3}$,
Andrei Zinovyev\,$^{1,2,3}$%
\footnote{To whom correspondence should be addressed.
Tel: +33 (0)1 56 24 69 89; Fax: +33 (0)1 56 24 69 11; Email: andrei.zinovyev@curie.fr}}

\address{%
$^{1}$Institut Curie, 26 rue d'Ulm, 75248 Paris, France,
$^{2}$INSERM U900, 75248 Paris, France,
$^{3}$Mines ParisTech, 77300 Fontainebleau, France,
$^{4}$Sysra, 91330 Yerres, France.
}
% Affiliation must include:
% Department name, institution name, full road and district address,
% state, Zip or postal code, country

\history{%
Received January 1, 2009;
Revised February 1, 2009;
Accepted March 1, 2009}


\bibliographystyle{nar}
\maketitle

\begin{abstract}
NaviCell is a web tool using Google Maps API for easy and intuitive navigation
of large maps of molecular interactions. Here, we present the NaviCell Web
Service for network-based visualization of  ``omics" data, a major extension of
NaviCell current functionalities.  NaviCell Web Service implements a number of
data visual representation methods, including the novel map staining technique
which allows grasping large-scale trends in numerical values (such as whole
transcriptome) projected on top of the pathway map. The service also provides a
server mode, which allows to automate visualization tasks and retrieve data
from maps via RESTfull (standard HTTP) calls.  For convenience, complete
bindings to the Python programming languages are provided, with upcoming
bindings to other languages (R, Java).  We illustrate the interest of the tool
with  two case studies using pathway maps created by different research groups,
in which data visualization serves for getting new insights into molecular
mechanisms involved in systemic disease progression (cancer and
neurodegenerative disease).

\end{abstract}


\section{Introduction}

Biology is a scientific discipline deeply grounded in visual representations
serving for communicating results and ideas. Nowadays, there is a strong
incentive to provide \emph{interactive} web-based visual representations, where
users can easily play with different options to tweak and adapt them. Modern
molecular biology is particularly demanding for development of new tools
representing in a meaningfull way numerous ``omics" data collected in
large-scale experiments. Visualizing quantitative ``omics" data in the context
of biological networks provides insights into the molecular mechanisms in
healthy tissues and disease \cite{Gehlenborg2010,Barillot2012}. This is one of
the most demanded features of existing pathway databases such as KEGG 
and Reactome \cite{Kanehisa2012, Croft2014}. To answer this need, many tools
have been developed allowing mapping ``omics" data on top of the biological
networks \cite{Arakawa2005,vanIersel2008,Luo2013, Nishida2014}. These tools use
the content of existing pathway databases to retrieve the pathway information
through dedicated APIs. In addition, some pathway databases provide graphical
user interfaces (GUI) to perform the task of data visualization, on top of
their interactive pathway maps.

When a user is faced with the necessity to visualize the ``omics" data on top of
biological networks, there are currently two options available: either to use
GUI of a suitable pathway database, which can be a tedious manual work, or use
one of the available standalone or web-based tools for creating static images
of colored pathways which are not interactive anymore and do not allow browsing
the molecular interactions. Currently, there exist no well-developed APIs
that allow to programmatically apply data visualization on top of biological
networks such that the omics data can be browsed simultaneously with the
pathway information.

Available network-based methods for data visualization have several common
limitations. First, most of them do not provide possibility of data
abstraction, allowing to visualize coarse-grained trends in the omics data.
This is needed if, for example, there is a wish to visualize the whole
transcriptome of a cell or a group of samples on top of a large network,
representing a big part of cellular interactome. Current methods usually attach
some elements of standard scientific graphics (gradient color, heatmaps,
barplots) to the individual elements of pathway maps, which makes the
visualization hardly readable at higher levels of zoom as well as the map
content itself. Second, most of data visualization tools are specific to a
particular pathway database structure, format and the way of graphical pathway
representation. This limits the use of data visualization on top of
user-defined pathway maps. Third, APIs for programmatic web-based data
visualization are practically non-existent.

To fullfill this need, we have developed NaviCell Web Service tool. It allows
visualizing omics data via GUI and flexible API using standard and advanced
methods of data visualization including a possibility to perceive the data
structure at different network scales. The NaviCell Web Service allows
exploiting pathway maps created by users themselves or from existing databases,
including very large maps (for example, representing the content of a whole
pathway database).


% **************************************************************
% Keep this command to avoid text of first page running into the
% first page footnotes
\enlargethispage{-65.1pt}
% **************************************************************





The NaviCell Web Service for Network-based Data
Visualization is a combination of 1) a user-friendly NaviCell JavaScript-based
web interface \cite{kuperstein2013navicell}, allowing browsing very large maps
of biological networks using Google Maps API using \emph{semantic zooming}
principle and visualize ``omics" data on top of them
(\href{https://navicell.curie.fr})); 2) a large collection of maps available for
data visualization including the Atlas of Cancer Signaling Network (ACSN)
(\href{https://acsn.curie.fr}))  and other detailed network maps created by many
research groups for a particular subject (Toll-like receptor signalling, EGFR
and mTOR pathways, mast cell activation, dendritic cells, iron metabolism,
Alzheimer disease and others); 3) a REST API allowing programmatic use of all
data visualization functions and manipulating the web interface, with available
Python binding and upcoming R and Java bindings.

A quick live example which \emph{automatically} uploads a sample dataset from
TCGA database and illustrates all basic data visualization capabilities of
NaviCell Web Service is provided at
\href{http://navicell.curie.fr/pages/nav\_web\_service.html}.. Sample Python
code for quick start on using NaviCell Web Service API is available from the
same page in the ``Python API and data files" section.

\section{ALGORITHMS AND SOFTWARE}

\subsection{Implementation}

We conceived the NaviCell Web Service for Network-based Data Visualization
as a significant development of NaviCell tool \cite{kuperstein2013navicell} allowing
users to visualize and analyze different types of ``omics" data. The web service
is implemented at two interconnected functional levels.
The first level corresponds to the interactive dialog-based
use of the web interface to load and visualize data. This level is implemented
as a specific set of JavaScript functions linked to the menu located
in the right-hand panel of the NaviCell web interface. The second level corresponds to a
programmatic usage of the service, which allows users to write the code that will
communicate with the NaviCell \emph{server} in order to automate all the
visualization operations. This functional level is implemented as a RESTful web
service, using standard web protocol HTTP operations and data encoding (JSON)
to perform all the necessary operations \cite{fielding2002principled}.
RESTful web services are very popular and have the advantage of being
lightweight, efficient and simple to use. They also have the benefit of
relatively ease of implementation in many different programming languages and
software packages. At the moment we have a full API ready for Python, and
APIs for R and Java in development. These APIs will facilitate the
integration of NaviCell Web Service in other applications: among them, there exists an ongoing work on providing
the data visualization services through the Garuda Alliance web platform, which aims to be a one-stop service for
bioinformatics and systems biology \cite{ghosh2011software} and GeneSpring software \cite{Chu2001}. Figure 1 is
summarizing the NaviCell \emph{server} architecture and information flow.
\subsection{Input data types}

The pathway maps which can be used for NaviCell Web Serivce data visualization
are to be prepared using CellDesigner tool\cite{Funahashi2008}, which is widely used in systems biology
community and implements the Systems Biology Graphical Notation standard \cite{LeNovere2009}
for representing biological networks. These pathway maps are converted into
interactive Google Maps-based web-interfaces via NaviCell tool\cite{kuperstein2013navicell}.

NaviCell data visualization web service is able to process
several types of ``omics" data. The complete list of data types that are
currently accepted for input is provided in Table \ref{table:01}. The different biological data
types are mapped into various internal representations that determine what methods
of data visualization can be applied.

For instance, a mRNA expression data matrix is associated to a
``continuous" numerical internal representation. Thus, if the user is choosing to display
this data with a heatmap, mapping to a color gradient will applied by default, with a possibility
for modifying the default setting. On the other hand, when a
matrix with discrete copy-number data is loaded, it is associated with a
``discrete ordered" internal representation for which
a specific color palette is applied for visualization,
with a distinct color associated to each discrete copy-number state.

The input format for data sets is standard tab-delimited text files, with rows
representing genes or their products and columns representing samples (or experiments, or time
points). Genes in the first column should be labeled by their standard HUGO (HGNC)
gene symbols, that will be associated to the different entities (genes,
proteins, complexes) on the pathway map.

Users can also upload sample annotation files to specify how samples can organized into meaningfull groups
(e.g., disease vs control), as a simple tab-delimited text file with sample names in
rows and annotation fields as columns. Then, an appropriate method will be used
to summarize the values of all the samples contained in a group, defined by the
internal data representation. For instance, expression values for a group of samples
is averaged by default for the ``continuous" data type, but the user will have
a possibility to change the default method and use the median or
the maximum value in the group, etc. For mutation data (``discrete unordered" internal data type),
taking the average does not make sense and instead the grouping method by default is
``at least one element of the group is mutated".

%\begin{table}[b]
\begin{table}
\tableparts{%
\caption{List of input biological data types for NaviCell Web Service data visualization. The
first column lists the types as they appear in the NaviCell Web Service interface. The
second column lists the internal data representations that are used to determine what
type of data visualization can be applied for a selected data type.}
\label{table:01}%
}{%
\begin{tabular*}{\columnwidth}{@{}ll@{}}
\toprule
``Omics" data type &  Internal representation
%\\
%& (\%) & (s$^{-1}$) & (\%) & (s$^{-1}$)
\\
\colrule
mRNA expression data & continuous
\\
microRNA expression data & continuous
\\
Protein expression data & continuous
\\
Discrete copy number data & discrete ordered
\\
Continuous copy number data & continuous
\\
Mutation data & discrete unordered
\\
Gene list & set
\\
\botrule
\end{tabular*}%
}
{
}
\end{table}

\subsection{Graphical data representations}

We have included in the NaviCell Web Service for network-based data visualization several methods
for graphical representation of molecular data. Some of them are standard
and broadly used in molecular biology (heatmaps and barplots), while others (glyphs and map
staining) have not been previously employed (to our knowledge) in the context of network-based
data visualization.

\begin{itemize}

\item \textbf{Simple markers} are pictograms (similar to the ones used in Google Maps
to indicate the geographical locations) are drawn on the pathway map at the location
of a given molecular species (protein, gene, complex, phenotype). They are used to display the
results of a search performed by the user or to display a list of genes of interest
uploaded on the map.

\item \textbf{Heatmaps} display individual values of a data matrix as
colors. They are often used in molecular biology, in particular, for displaying expression
data. In NaviCell Web Service, heatmaps can be used to
display continuous data types such as expression values, or discrete data such
as copy-number or mutation values. The user can arrange the display to have
several samples or group of samples displayed in several data sets
(for example, for showing simultaneously, gene promoter methylation and expression values).

\item \textbf{Barplots} are charts with rectangular bars proportional to the values
they represent. On NaviCell Web Sevice interface, they are colored according to the
data value. Barplots can be efficient to visually distinguish numerical
values between two or more groups of conditions (e.g. disease vs control).

\item \textbf{Glyphs} are graphical representations using basic geometrical
shapes (triangle, square, rectangle, diamond, hexagon). The user can specify
which datasets to use to map on the shape, the size and the color of the glyph. This
type of representation is particularly useful to visually combine different
types of data. For example, one might consider using matched data for
gene copy number and expression. In this case, the shape of the
glyphs could be assigned to the copy-number values, while the expression data
would be used for the color of the glyph. Like this, users can quickly
appreciate the two data types values at the same time on the glyph.

\item \textbf{Map staining} is a novel network-based data visualization method where
background areas (or territories) around each molecular entity are colored according to the data
value associated to this entity. The area occupied by a particular map element is defined as the Voronoi's cell
associated to this element. The Voronoi's cell is a convex polygon containing all the points
of the territory which are closer to a chosen element than to any other element
\cite{aurenhammer1991voronoi}. In our case, we use all the entities present on a molecular map
for defining the Voronoi's cells. The polygons are pre-computed for each map and post-processed
to avoid too large polygons, and are colored according to the data values and the options
defined by the user. In map staining, the colorful decorations of the pathway map
are replaced by a black-and-white background in order to avoid mixing the colors.
Using map staining allows to appreciate large-scale trends in molecular data mapped
onto biological networks, when observed ``from far" (see example in Figure 2).

\end{itemize}



\section{RESULTS}

We show that NaviCell Web Service can be used for simultaneous analysis of
different types of high-throughput data in two case studies:
\begin{enumerate}
\item comparing two prostate cancer cell lines using the cell cycle pathway map
and transcriptomics and mutation data.

\item using the map of molecular interactions involved in Alzheimer disease
\cite{Mizuno2012} for visualizing the transcriptome data collected for
different brain areas \cite{Hokama2014}.
\end{enumerate}

Furthermore, several examples of using NaviCell Web Service for data
visualization can be found in NaviCell Web Service user guide and case studies
provided at \href{http://navicell.curie.fr/pages/nav\_web\_service.html}.. For
instance, we have mapped ovary cancer data downloaded from The Cancer Genome
Atlas \cite{TCGA2011Ovarian} on a  large map of Atlas of Cancer Signalling
Network \cite{Kuperstein2015}, and a complete description of this analysis is
available as supplementary material. 


\subsection{Comparing transcriptomes of prostate cancer cell lines}

As an illustration of the use of maps in comparing two cancer genomic profiles, we
selected two prostate cancer cell lines from the Cancer Cell Line Encyclopedia
\cite{barretina2012cancer}: a prostate hormone-sensitive
tumor cell line (LNCAP), and a prostate hormone-resistant tumor cell line
(DU145). We gathered gene expression, copy number and mutation data for
these two cell lines and mapped them onto the cell cycle map
\cite{calzone2008comprehensive}.

If we consider the cell line data as the mean expression of genes of a
population of asynchronous cells, most LNCAP cells seem to be expressing genes
from the early state G1 (Figure2A, upper-right area) and the G1-S checkpoint
(Figure2A, lower-left area) while most DU145 cells express genes from the
later stages of the cell cycle (Figure 2a, G1-late, S-phase and G2-phase
areas).  We notice that relatively few cell cycle genes are mutated, amplified or lost
in these two cell lines (Figureg2A and 2B, square and triangle glyphs).

By zooming on the map, we can determine that for the LNCAP cells, the most
important gene alterations are: the amplification of DP1 (present in most
complexes involving E2F1), and the homozygous loss of E2F2, transcription
factor. The mutations concern mainly genes from the apoptotic pathway: ATR or
CHEK2. As for the expression, the more noticeable trend is that the expression
of the cell cycle inhibitors, such as RBL2, p21, CDC25C, is still high, whereas
the expression of the G2 cyclin, for instance, CyclinA is low.

For the DU145 cells, genes involved in later stages of the cell cycle seem to
be more expressed compared to the LNCAP cells. It can be that if the cells were
arrested in LNCAP, they are more advanced in the cycle in DU145 cells by
overpassing the G1/S checkpoint and are arrested at the spindle checkpoint.
They are more prone to proliferate than the LNCAP cells. The expression of some
cyclins seems to confirm this fact: CyclinB, CyclinD and CyclinH are higher
than in LNCAP. Some means to stop the cycle seem to be kept though, with Cdc20
and Cdc25 expression high.

We checked the expression levels of KI-67, a marker of proliferation, in the
dataset and its expression is indeed higher in DU145 than in LNCAP cells, which
tends to confirm our hypothesis that DU145 are more proliferative than LNCAP
cells.

\subsection{Creating molecular portrait of Alzheimer's disease}

[Inna's contribution]

\section{DISCUSSION}

We have compared the features of the NaviCell Web Service for data visualization
with similar tools. We selected a number of web-based tools that are providing
similar functionalities, i.e. easy pathway browsing functions and interactive data
visualization capabilities. For the comparison, we have focused on the features
related to map navigation, molecular data types, graphical representations,
and programmatic access. Table \ref{table:02} display the results. NaviCell Web
Service offers more features than other tools in terms of map navigation, for
data visualization with additional graphical representations (such as map
staining) and more data types, and finally extended support for programmatic
access from different computer languages.

To extend the scope of NaviCell Web Service, we are currently working on porting some of the
most used pathway databases into the CellDesigner format which can be used after for displaying
in NaviCell, with a possibility of data visualization.


%\begin{table}[b]
\begin{table}
\tableparts{%
\caption{Comparison of the NaviCell Web Service features with similar web sites for pathway-based data visualization.
}
\label{table:02}%
}{%
\begin{tabular*}{\columnwidth}{@{}lcccccc@{}}
\toprule
 Features & Na & Re & KE & iP & Bc & Pa
%\\
%& (\%) & (s$^{-1}$) & (\%) & (s$^{-1}$)
\\
\colrule
Map: navigation & $\bullet$ & $\bullet$ & & $\bullet$ & $\bullet$ & $\bullet$
\\
Map: simple zooming & $\bullet$ & $\bullet$ & $\bullet$ & $\bullet$ & $\bullet$ &  $\bullet$
\\
Map: semantic zooming & $\bullet$ & & & & &
\\
Visualization: node coloring & & $\bullet$ & & & & $\bullet$
\\
Visualization: heatmaps & $\bullet$ & & & & $\bullet$ &
\\
Visualization: barplots & $\bullet$ & & & & $\bullet$ &
\\
Visualization: glyphs & $\bullet$ & & & & &
\\
Visualization: map staining & $\bullet$ & & & & &
\\
Data mapping: gene lists& $\bullet$ & $\bullet$ & $\bullet$ & $\bullet$ & $\bullet$ & $\bullet$
\\
Data mapping: expression data& $\bullet$ & $\bullet$ & & & $\bullet$ & $\bullet$
\\
Data mapping: copy-number data& $\bullet$ & & & & &
\\
Data mapping: mutation data& $\bullet$ & & & & &
\\
Data mapping: metabolomic data& & $\bullet$ & & & &
\\
Data mapping: interactions & & $\bullet$ & & & &
\\
Programmatic access: RESTful web & $\bullet$ & $\bullet$ & $\bullet$ & & &
\\
Programmatic access: data visual. & $\bullet$ & & & & &
\\
\botrule
\end{tabular*}%
}
{Abbreviations: Na: NaviCell Web Service, Re: Reactome\cite{Croft2014},
KE: KEGG\cite{Kanehisa2012}, iP: iPath\cite{letunic2008ipath}, Bc:
BioCyc\cite{karp2005expansion}, Pa: PATIKAweb\cite{demir2002patika}.
}
\end{table}


\section{CONCLUSION}

NaviCell Web Service should contribute to the growing set of highly demanded tools for molecular biology allowing visualization of ``omics" data in the context of biological network maps.

\section{ACKNOWLEDGEMENTS}

Agilent. PIC SysBio. INVADE. COMET.

\subsubsection{Conflict of interest statement.} None declared.
\newpage

\bibliography{biblio}

\section{FIGURE LEGENDS}

\textbf{Figure 1.} General architecture of the NaviCell Web service
\emph{server}. Client software (light blue layer) communicates with the server
(red layer) through standard HTTP requests using the standard JSON format to
encode data (RESTful web service, dark blue layer). A session (with a unique ID)
is established between the server and the client browser (yellow layer) through
Ajax communication channel to visualize the results of the commands send by the
client. It is worth noticing that communication channels are bidirectional, i.e.
the client software can send data (e.g. an expression data matrix) to the
server, but it can also receive data from the server (e.g. a list of gene HUGO
codes contained in a map).


\end{document}
